common/management/src/role/role_mgr.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_exception::ToErrorCode;
+use common_meta_api::KVApi;
+use common_meta_types::GrantObject;
+use common_meta_types::IntoSeqV;
+use common_meta_types::MatchSeq;
+use common_meta_types::MatchSeqExt;
+use common_meta_types::OkOrExist;
+use common_meta_types::Operation;
+use common_meta_types::RoleInfo;
+use common_meta_types::SeqV;
+use common_meta_types::UpsertKVAction;
+use common_meta_types::UserPrivilegeSet;
+
+use crate::role::role_api::RoleMgrApi;
+
+static ROLE_API_KEY_PREFIX: &str = "__fd_roles";
+
+pub struct RoleMgr {
+    kv_api: Arc<dyn KVApi>,
+    role_prefix: String,
+}
+
+impl RoleMgr {
+    pub fn new(kv_api: Arc<dyn KVApi>, tenant: &str) -> Self {
+        RoleMgr {
+            kv_api,
+            role_prefix: format!("{}/{}", ROLE_API_KEY_PREFIX, tenant),
+        }
+    }
+
+    async fn upsert_role_info(
+        &self,
+        role_info: &RoleInfo,
+        seq: Option<u64>,
+    ) -> common_exception::Result<u64> {
+        let key = format!("{}/{}", self.role_prefix, &role_info.name);
+        let value = serde_json::to_vec(&role_info)?;
+
+        let match_seq = match seq {
+            None => MatchSeq::GE(1),
```

Comment:
If you allow setting value on an absent record, `MatchSeq::Any` may be better: `GE(1)` means a replace operation, which requires the record has to be present.

Since the method name is `upsert` I suppose it allows to `add` a record.
Otherwise, `update_xxx` would be better.